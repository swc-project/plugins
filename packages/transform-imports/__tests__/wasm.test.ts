import {expect, test, describe} from "vitest";
import {transform} from "@swc/core";
import path from "node:path";
import url from "node:url";
import fs from "node:fs/promises";

const pluginName = "swc_plugin_transform_imports.wasm";

const transformCode = async (
    code: string,
    options: Record<string, unknown> = {
        "react-bootstrap": {
            transform: "react-bootstrap/lib/{{member}}",
            preventFullImport: false,
            skipDefaultConversion: false,
            handleDefaultImport: true,
            handleNamespaceImport: true,
        },
        "my-library/?(((\\w*)?/?)*)": {
            transform: "my-library/{{ matches.[1] }}/{{member}}",
            preventFullImport: false,
            skipDefaultConversion: false,
            handleDefaultImport: true,
            handleNamespaceImport: true,
        },

        "rewrite-module-namespace": {
            transform: "rewrite-module-namespace/{{ member }}",
            prevent_full_import: false,
            skip_default_conversion: true,
            handle_default_import: true,
            handle_namespace_import: true,
            rewrite_namespace_to_proxy: true
        },
        "my-library-2": {
            transform: "my-library-2/{{ camelCase member }}",
            preventFullImport: false,
            skipDefaultConversion: true,
            handleDefaultImport: true,
            handleNamespaceImport: true,
        },
        "my-library-4": {
            transform: [
                ["foo", "my-library-4/this_is_foo"],
                ["bar", "my-library-4/this_is_bar"],
                [
                    "use(\\w*)",
                    "my-library-4/{{ kebabCase member }}/{{ kebabCase memberMatches.[1] }}",
                ],
                ["(\\w*)Icon", "my-library-4/{{ kebabCase memberMatches.[1] }}"],
                ["*", "my-library-4/{{ kebabCase member }}"],
            ],
            preventFullImport: false,
            skipDefaultConversion: false,
            handleDefaultImport: true,
            handleNamespaceImport: true,
        },
        "my-library-3": {
            transform: "my-library-3/{{ kebabCase member }}",
            preventFullImport: false,
            skipDefaultConversion: true,
            handleDefaultImport: true,
            handleNamespaceImport: true,
        },
        "^(\\..*)(\\.tsx?)$": {
            transform: "{{matches.[1]}}.js",
            preventFullImport: false,
            skipDefaultConversion: false,
            handleDefaultImport: false,
            handleNamespaceImport: false,
        },
    },
) => {
    return transform(code, {
        jsc: {
            parser: {
                syntax: "ecmascript",
                jsx: true,
            },
            target: "es3",
            experimental: {
                plugins: [
                    [
                        path.join(
                            path.dirname(url.fileURLToPath(import.meta.url)),
                            "..",
                            pluginName,
                        ),
                        options,
                    ],
                ],
            },
        },
        filename: "test.js",
    });
};

async function walkDir(
    dir: URL,
    callback: (
        dir: string,
        input: string,
        config?: Record<string, unknown>,
    ) => Promise<void>,
) {
    const dirs = await fs.readdir(dir);
    const baseDir = url.fileURLToPath(dir);

    for (const dir of dirs) {
        const inputFilePath = path.join(baseDir, dir, "input.js");
        const configPath = path.join(baseDir, dir, "config.json");

        const config = await fs.readFile(configPath, "utf-8").then(
            (json) => {
                return JSON.parse(json);
            },
            (_) => undefined,
        );

        try {
            const input = await fs.readFile(inputFilePath, "utf-8");
            await callback(dir, input, config);
        } catch (e) {
            console.log(e);
        }
    }
}

describe("Should load transform-imports wasm plugin correctly", async () => {
    await walkDir(
        new URL("../transform/tests/fixture", import.meta.url),
        async (dir, input, config) => {
            await test(`Should transform ${dir} correctly`, async () => {
                const {code} = await transformCode(input, config);
                expect(code).toMatchSnapshot();
            });
        },
    );
});
