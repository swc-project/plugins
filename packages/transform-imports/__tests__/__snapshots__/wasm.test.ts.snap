// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should load transform-imports wasm plugin correctly > Should transform default correctly 1`] = `
"import defaultExport from "my-default";
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform dynamic-imports correctly 1`] = `
"await import("./stringLiteral.js");
await import("./templateLiteral.js");
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform mapping correctly 1`] = `
"import foo from "my-library-4/this_is_foo";
import bar from "my-library-4/this_is_bar";
import otherImport from "my-library-4/other-import";
import useSomeAPI from "my-library-4/use-some-api/some-api";
import ArrowUpIcon from "my-library-4/arrow-up";
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform mapping-export correctly 1`] = `
"export { default as foo } from "my-library-4/this_is_foo";
export { default as bar } from "my-library-4/this_is_bar";
export { default as otherImport } from "my-library-4/other-import";
export { default as useSomeAPI } from "my-library-4/use-some-api/some-api";
export { default as ArrowUpIcon } from "my-library-4/arrow-up";
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform mixed-default correctly 1`] = `
"import defaultExport, { someNamedExport } from "my-mixed-named";
import otherDefaultExport, * as starExport from "my-mixed-star";
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform namespace correctly 1`] = `
"import * as someModule from "my-module-namespace";
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform namespace-export correctly 1`] = `
"import * as _someModule from "my-module-namespace";
export { _someModule as someModule };
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform regex correctly 1`] = `
"import MyModule from "my-library/MyModule";
import App from "my-library/components/App";
import Header from "my-library/components/App/Header";
import Footer from "my-library/components/App/Footer";
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform regex-export correctly 1`] = `
"export { default as MyModule } from "my-library/MyModule";
export { default as App } from "my-library/components/App";
export { default as Header } from "my-library/components/App/Header";
export { default as Footer } from "my-library/components/App/Footer";
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform rewrite-namespace correctly 1`] = `
"import * as test from 'rewrite-module-namespace';
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform rewrite-namespace-export correctly 1`] = `
"import * as _someModule from "rewrite-module-namespace";
export { _someModule as someModule };
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform side-effect-imports correctly 1`] = `
"import './upload/upload.js';
import './scripts/scripts.model.js';
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform simple correctly 1`] = `
"import Grid from "react-bootstrap/lib/Grid";
import Row from "react-bootstrap/lib/Row";
import Col1 from "react-bootstrap/lib/Col";
import { MyModule } from "my-library-2/myModule";
import { Widget } from "my-library-2/widget";
import { MyModule } from "my-library-3/my-module";
"
`;

exports[`Should load transform-imports wasm plugin correctly > Should transform simple-export correctly 1`] = `
"export { default as Grid } from "react-bootstrap/lib/Grid";
export { default as Row } from "react-bootstrap/lib/Row";
export { default as Col1 } from "react-bootstrap/lib/Col1";
export { MyModule } from "my-library-2/myModule";
export { Widget } from "my-library-2/widget";
export { MyModule } from "my-library-3/my-module";
"
`;
